File: internal/agent/claude_session.go
// Copyright (c) Mehmet Bektas <mbektasgh@outlook.com>

package agent

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"os/exec"
	"sync"
	"syscall"
)

type persistentClaudeSession struct {
	mu sync.Mutex

	cmd    *exec.Cmd
	stdin  io.WriteCloser
	stdout *bufio.Scanner

	binaryPath   string
	model        string
	systemPrompt string
	sandbox      string
	useBwrap     bool
	workingDir   string

	claudeSessionID string
}

func (s *persistentClaudeSession) matches(agent *ClaudeCodeAgent, req RunRequest) bool {
	if s == nil || s.cmd == nil || !processAlive(s.cmd) {
		return false
	}
	model := agent.modelName()
	useBwrap := agent.useBwrap()
	return s.binaryPath == agent.binaryPath() &&
		s.model == model &&
		s.systemPrompt == req.SystemPrompt &&
		s.sandbox == agent.Sandbox &&
		s.useBwrap == useBwrap &&
		s.workingDir == req.WorkingDir
}

func processAlive(cmd *exec.Cmd) bool {
	if cmd == nil || cmd.Process == nil || cmd.ProcessState != nil {
		return false
	}
	return cmd.Process.Signal(syscall.Signal(0)) == nil
}

func (s *persistentClaudeSession) runTurn(ctx context.Context, agent *ClaudeCodeAgent, req RunRequest, events chan<- Event) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	if err := s.ensureProcess(ctx, agent, req); err != nil {
		return err
	}

	userMsg := map[string]any{
		"type": "user",
		"message": map[string]any{
			"role":    "user",
			"content": []map[string]string{{"type": "text", "text": req.Message}},
		},
	}
	sessionID := s.claudeSessionID
	if sessionID == "" {
		sessionID = req.SessionID
	}
	if sessionID != "" {
		userMsg["session_id"] = sessionID
	}

	payload, err := json.Marshal(userMsg)
	if err != nil {
		return err
	}
	payload = append(payload, '\n')

	if err := s.writePayload(payload); err != nil {
		if err2 := s.restart(ctx, agent, req); err2 != nil {
			return err2
		}
		if err := s.writePayload(payload); err != nil {
			return fmt.Errorf("write claude stdin after restart: %w", err)
		}
	}

	parser := newClaudeStreamParser()
	for {
		if err := ctx.Err(); err != nil {
			return err
		}
		if !s.stdout.Scan() {
			s.stopLocked()
			return fmt.Errorf("claude process ended unexpectedly")
		}

		line := s.stdout.Bytes()
		if len(line) == 0 {
			continue
		}

		var envelope struct {
			Type      string `json:"type"`
			SessionID string `json:"session_id"`
			IsError   bool   `json:"is_error"`
			Error     any    `json:"error"`
		}
		if err := json.Unmarshal(line, &envelope); err != nil {
			continue
		}

		parser.handleLine(line, events)

		switch envelope.Type {
		case "result":
			if !envelope.IsError && envelope.SessionID != "" {
				s.claudeSessionID = envelope.SessionID
			}
			return nil
		case "error":
			msg := formatClaudeError(envelope.Error)
			events <- Event{Type: EventError, Error: msg}
			return nil
		}
	}
}

func formatClaudeError(err any) string {
	switch v := err.(type) {
	case string:
		if v != "" {
			return v
		}
	case map[string]any:
		if msg, _ := v["message"].(string); msg != "" {
			return msg
		}
		b, _ := json.Marshal(v)
		if len(b) > 0 {
			return string(b)
		}
	}
	return "unknown error"
}

func (s *persistentClaudeSession) writePayload(payload []byte) error {
	if s.stdin == nil {
		return fmt.Errorf("claude stdin unavailable")
	}
	_, err := s.stdin.Write(payload)
	return err
}

func (s *persistentClaudeSession) ensureProcess(ctx context.Context, agent *ClaudeCodeAgent, req RunRequest) error {
	if s.matches(agent, req) {
		return nil
	}
	s.stopLocked()
	return s.start(ctx, agent, req)
}

func (s *persistentClaudeSession) restart(ctx context.Context, agent *ClaudeCodeAgent, req RunRequest) error {
	s.stopLocked()
	return s.start(ctx, agent, req)
}

func (s *persistentClaudeSession) start(ctx context.Context, agent *ClaudeCodeAgent, req RunRequest) error {
	bin := agent.binaryPath()
	model := agent.modelName()
	useBwrap := agent.useBwrap()

	if req.SessionID != "" && s.claudeSessionID == "" {
		s.claudeSessionID = req.SessionID
	}

	args := []string{
		"-p",
		"--input-format", "stream-json",
		"--output-format", "stream-json",
		"--verbose",
		"--dangerously-skip-permissions",
		"--include-partial-messages",
		"--model", model,
	}
	if req.SystemPrompt != "" {
		args = append(args, "--system-prompt", req.SystemPrompt)
	}
	resume := s.claudeSessionID
	if resume == "" {
		resume = req.SessionID
	}
	if resume != "" {
		args = append(args, "--resume", resume)
	}

	var cmd *exec.Cmd
	if useBwrap {
		bwrap := GetBwrapStatus()
		wrappedBin, wrappedArgs := WrapWithBwrap(bwrap.Path, bin, args, req.WorkingDir)
		cmd = exec.CommandContext(ctx, wrappedBin, wrappedArgs...)
	} else {
		cmd = exec.CommandContext(ctx, bin, args...)
		if req.WorkingDir != "" {
			cmd.Dir = req.WorkingDir
		}
	}

	stdin, err := cmd.StdinPipe()
	if err != nil {
		return err
	}
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return err
	}
	stderr, err := cmd.StderrPipe()
	if err != nil {
		return err
	}

	if err := cmd.Start(); err != nil {
		fmt.Fprintf(os.Stderr, "[claude] start error: %v\n", err)
		return err
	}

	go func() {
		scanner := bufio.NewScanner(stderr)
		for scanner.Scan() {
			fmt.Fprintf(os.Stderr, "[claude stderr] %s\n", scanner.Text())
		}
	}()

	sc := bufio.NewScanner(stdout)
	sc.Buffer(make([]byte, 4*1024*1024), 4*1024*1024)

	s.cmd = cmd
	s.stdin = stdin
	s.stdout = sc
	s.binaryPath = bin
	s.model = model
	s.systemPrompt = req.SystemPrompt
	s.sandbox = agent.Sandbox
	s.useBwrap = useBwrap
	s.workingDir = req.WorkingDir
	return nil
}

func (s *persistentClaudeSession) stop() {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.stopLocked()
}

func (s *persistentClaudeSession) stopLocked() {
	if s.stdin != nil {
		s.stdin.Close()
		s.stdin = nil
	}
	if s.cmd != nil && s.cmd.Process != nil && s.cmd.ProcessState == nil {
		s.cmd.Process.Kill()
		_, _ = s.cmd.Process.Wait()
	}
	s.cmd = nil
	s.stdout = nil
}

File: docker/claude_session.py
"""Persistent Claude CLI session using the stream-json stdin/stdout protocol."""

from __future__ import annotations

import json
import os
import subprocess
import sys
import threading
from subprocess import PIPE
from typing import Any, Callable, Generator

from claude_stream import new_claude_stream_parser
from harness_env import subprocess_env


WrapArgsFn = Callable[[list[str], str], list[str]]


def run_ephemeral_claude_turn(
    message: str,
    working_dir: str,
    model: str = "",
    system_prompt: str = "",
    wrap_args: WrapArgsFn | None = None,
    **kwargs: Any,
) -> Generator[dict[str, Any], None, None]:
    """One-shot Claude turn that never resumes an existing session or mutates a persistent session."""
    wrap = wrap_args or (lambda args, _wd: args)
    wd = working_dir or os.getcwd()

    claude_args = [
        "claude",
        "-p",
        "--input-format",
        "stream-json",
        "--output-format",
        "stream-json",
        "--verbose",
        "--dangerously-skip-permissions",
        "--include-partial-messages",
    ]
    if model:
        claude_args += ["--model", model]
    if system_prompt:
        claude_args += ["--system-prompt", system_prompt]
    if kwargs.get("userScopeHarness"):
        claude_args += ["--setting-sources", "user,project,local"]
        mcp_path = os.path.join("/home/nui/.nui/session-config", ".claude.json")
        if os.path.isfile(mcp_path):
            claude_args += ["--mcp-config", mcp_path]

    args = wrap(claude_args, wd)
    cwd = None if os.environ.get("NUI_BWRAP_PATH") else wd

    proc = subprocess.Popen(
        args,
        stdin=PIPE,
        stdout=PIPE,
        stderr=PIPE,
        cwd=cwd,
        env=subprocess_env(kwargs),
        start_new_session=True,
    )
    assert proc.stdin is not None and proc.stdout is not None

    def drain_stderr() -> None:
        assert proc.stderr is not None
        for line in proc.stderr:
            print(f"[claude stderr] {line.decode(errors='replace').rstrip()}", file=sys.stderr, flush=True)

    threading.Thread(target=drain_stderr, daemon=True).start()

    user_msg: dict[str, Any] = {
        "type": "user",
        "message": {
            "role": "user",
            "content": [{"type": "text", "text": message}],
        },
    }
    payload = (json.dumps(user_msg) + "\n").encode()
    try:
        proc.stdin.write(payload)
        proc.stdin.flush()
        proc.stdin.close()
    except (BrokenPipeError, OSError):
        proc.kill()
        yield {"type": "error", "error": "claude process ended unexpectedly"}
        return

    parser = new_claude_stream_parser()
    while True:
        raw = proc.stdout.readline()
        if not raw:
            proc.wait(timeout=5)
            yield {"type": "error", "error": "claude process ended unexpectedly"}
            return

        line = raw.decode(errors="replace").strip()
        if not line:
            continue

        try:
            envelope = json.loads(line)
        except json.JSONDecodeError:
            continue

        yield from parser.handle_envelope(envelope)

        msg_type = envelope.get("type")
        if msg_type == "result":
            if envelope.get("is_error"):
                err = envelope.get("error")
                if isinstance(err, dict):
                    err = err.get("message") or json.dumps(err)
                yield {"type": "error", "error": str(err or "unknown error")}
            proc.wait(timeout=5)
            return

        if msg_type == "error":
            err = envelope.get("error")
            if isinstance(err, dict):
                err = err.get("message") or json.dumps(err)
            yield {"type": "error", "error": str(err or "unknown error")}
            proc.wait(timeout=5)
            return


class PersistentClaudeSession:
    """Keep one Claude CLI process alive and send each prompt over stdin."""

    def __init__(self, wrap_args: WrapArgsFn | None = None) -> None:
        self._wrap_args = wrap_args or (lambda args, _wd: args)
        self._lock = threading.Lock()
        self._proc: subprocess.Popen[bytes] | None = None
        self._stderr_thread: threading.Thread | None = None
        self._working_dir = ""
        self._model = ""
        self._system_prompt = ""
        self._claude_session_id = ""

    def run_turn(
        self,
        message: str,
        working_dir: str,
        resume_session_id: str = "",
        model: str = "",
        system_prompt: str = "",
        **kwargs: Any,
    ) -> Generator[dict[str, Any], None, None]:
        with self._lock:
            yield from self._run_turn_locked(
                message, working_dir, resume_session_id, model, system_prompt, kwargs,
            )

    def stop(self) -> None:
        with self._lock:
            self._stop_unlocked()

    def _run_turn_locked(
        self,
        message: str,
        working_dir: str,
        resume_session_id: str,
        model: str,
        system_prompt: str,
        kwargs: dict[str, Any],
    ) -> Generator[dict[str, Any], None, None]:
        self._ensure_process(working_dir, resume_session_id, model, system_prompt, kwargs)
        assert self._proc is not None and self._proc.stdin is not None and self._proc.stdout is not None

        user_msg: dict[str, Any] = {
            "type": "user",
            "message": {
                "role": "user",
                "content": [{"type": "text", "text": message}],
            },
        }
        session_id = self._claude_session_id or resume_session_id
        if session_id:
            user_msg["session_id"] = session_id

        payload = (json.dumps(user_msg) + "\n").encode()
        try:
            self._proc.stdin.write(payload)
            self._proc.stdin.flush()
        except (BrokenPipeError, OSError):
            self._stop_unlocked()
            self._ensure_process(working_dir, resume_session_id, model, system_prompt, kwargs)
            assert self._proc is not None and self._proc.stdin is not None and self._proc.stdout is not None
            self._proc.stdin.write(payload)
            self._proc.stdin.flush()

        parser = new_claude_stream_parser()
        while True:
            raw = self._proc.stdout.readline()
            if not raw:
                self._stop_unlocked()
                yield {"type": "error", "error": "claude process ended unexpectedly"}
                return

            line = raw.decode(errors="replace").strip()
            if not line:
                continue

            try:
                envelope = json.loads(line)
            except json.JSONDecodeError:
                continue

            yield from parser.handle_envelope(envelope)

            msg_type = envelope.get("type")
            if msg_type == "result":
                if not envelope.get("is_error"):
                    sid = envelope.get("session_id") or ""
                    if sid:
                        self._claude_session_id = sid
                return

            if msg_type == "error":
                err = envelope.get("error")
                if isinstance(err, dict):
                    err = err.get("message") or json.dumps(err)
                yield {"type": "error", "error": str(err or "unknown error")}
                return

    def _ensure_process(
        self,
        working_dir: str,
        resume_session_id: str,
        model: str,
        system_prompt: str,
        kwargs: dict[str, Any],
    ) -> None:
        wd = working_dir or os.getcwd()
        if (
            self._proc is not None
            and self._proc.poll() is None
            and self._working_dir == wd
            and self._model == model
            and self._system_prompt == system_prompt
        ):
            return

        self._stop_unlocked()
        if resume_session_id and not self._claude_session_id:
            self._claude_session_id = resume_session_id

        claude_args = [
            "claude",
            "-p",
            "--input-format",
            "stream-json",
            "--output-format",
            "stream-json",
            "--verbose",
            "--dangerously-skip-permissions",
            "--include-partial-messages",
        ]
        if model:
            claude_args += ["--model", model]
        if system_prompt:
            claude_args += ["--system-prompt", system_prompt]
        resume = self._claude_session_id or resume_session_id
        if resume:
            claude_args += ["--resume", resume]
        if kwargs.get("userScopeHarness"):
            claude_args += ["--setting-sources", "user,project,local"]
            mcp_path = os.path.join("/home/nui/.nui/session-config", ".claude.json")
            if os.path.isfile(mcp_path):
                claude_args += ["--mcp-config", mcp_path]

        args = self._wrap_args(claude_args, wd)
        cwd = None if os.environ.get("NUI_BWRAP_PATH") else wd

        proc = subprocess.Popen(
            args,
            stdin=PIPE,
            stdout=PIPE,
            stderr=PIPE,
            cwd=cwd,
            env=subprocess_env(kwargs),
            start_new_session=True,
        )
        self._proc = proc
        self._working_dir = wd
        self._model = model
        self._system_prompt = system_prompt

        def drain_stderr() -> None:
            assert proc.stderr is not None
            for line in proc.stderr:
                print(f"[claude stderr] {line.decode(errors='replace').rstrip()}", file=sys.stderr, flush=True)

        self._stderr_thread = threading.Thread(target=drain_stderr, daemon=True)
        self._stderr_thread.start()

    def _stop_unlocked(self) -> None:
        proc = self._proc
        self._proc = None
        self._working_dir = ""
        if proc is None:
            return
        if proc.poll() is None:
            proc.terminate()
            try:
                proc.wait(timeout=5)
            except subprocess.TimeoutExpired:
                proc.kill()
                proc.wait(timeout=2)
